hypergan/losses/base_loss.py

```diff
@@ -93,19 +93,21 @@ def half(tens, second):
     def gradient_penalty(self):
         config = self.config
         gan = self.gan
+        ops = self.gan.ops
         gradient_penalty = config.gradient_penalty
-        if has_attr(gan.inputs, 'gradient_penalty_label'):
-            x = gan.inputs.gradient_penalty_label
-        else:
-            x = gan.inputs.x
+        x = gan.inputs.x
         generator = self.generator or gan.generator
         g = generator.sample
         discriminator = self.discriminator or gan.discriminator
         shape = [1 for t in g.get_shape()]
         shape[0] = gan.batch_size()
         uniform_noise = tf.random_uniform(shape=shape,minval=0.,maxval=1.)
         print("[gradient penalty] applying x:", x, "g:", g, "noise:", uniform_noise)
-        interpolates = x + uniform_noise * (g - x)
+        if config.gradient_penalty_type == 'dragan':
+            mean, variance = tf.nn.moments(x, axes=[0, 1, 2, 3])
+            interpolates = x + uniform_noise * 0.5 * variance * tf.random_uniform(shape=ops.shape(x), minval=0.,maxval=1.)
```

Comment:
A couple of open questions:
* Should we bound this to [-1, 1] ?
* This equation is `uniform_noise` (values 0-1) * `variance` * 0.5 * `alpha`(0-1) + x.  Should one of those be values -1 to 1?

We may want to try gaussians, maybe even do some fun stuff with the tensorflow image processing utils https://www.tensorflow.org/api_guides/python/image  (Image Adjustments).

We can also render `interpolates` in a sampler.  I wonder if `nearby_image` would be a better name.